### PR TITLE
Replace x/net/websocket with gorilla/websocket.

### PR DESCRIPTION
**Pull request by howbazaar**
There was a tech-board decision some time in the last two years when we were discussing websocket libraries and why we had two. The decision then was to standardise on gorilla/websocket, but nothing happened.

I'm hoping that using the more standards compliant websocket library will help with some of the leaks we are seeing. Even if they don't help right now, the new library gives the client code access to the websocket ping/pong frames, which the old library didn't. This should allow us to have better recognition when the other end goes away.

## QA steps

Deploy an old controller, make sure there are workload agents. Upgrade the controller and look at the logs. All older clients, CLI and agents should continue to work as normal. New juju CLI should work with older controller.

## Documentation changes

No user impact at all.

### CODE REVIEW

File: rpc/server.go [hunk 1]
```diff
@@ -341,13 +341,15 @@ func (conn *Conn) input() {
 	close(conn.dead)
 }
 
+var errRemoteClosed = errors.New("remote closed connection")
```
Comment (wallyworld) [inline rpc/server.go]: appears unused?

File: apiserver/websocket.go [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+		logger.Errorf("problem initiating websocket: %v", err)
+		return
+	}
+	handler(conn)
+}
+
+// sendInitialErrorV0 writes out the error as a params.ErrorResult serialized
+// with JSON with a new line character at the end.
+//
+// This is a hangover from the initial debug-log streaming endoing where the
```
Comment (wallyworld) [inline apiserver/websocket.go]: Can we add a TODO here

File: rpc/server.go (hunk 1 above)
Comment (howbazaar) [inline rpc/server.go]: Fixed.

**Comment by howbazaar** [comment]
$$merge$$

**Pull request merged**